Dvl revamp

Williangalvani opened the pull request:
Rewrite frontend in vue+vuetify
Fix bugs:
 - origin now is only set once, but user can change the rov position with position_estimate messages
 - update to use new Waterlinked interface/api
 - fix mavlink2rest paths for latest release
 - use waterlinked-dvl.local as a default hostname
 - set rangefinder parameters automatically
 - don't store origin in settings file

To-do:
 - [x] proper logging
 - [x] make sure the set_origin/set_positiion thing is handled correctly
 - [ ] ~gyro calibration?~
 - [ ] ~dead reckoning reset?~

Fixes https://github.com/bluerobotics/BlueOS-Water-Linked-DVL/issues/5
Fixes #6
Fixes #8

![image](https://user-images.githubusercontent.com/4013804/188223972-6c512dc2-2c63-46a3-8f84-83045b379cc5.png)

ES-Alexander committed:
dvl-a50: mavlink2resthelper: fix old paths

Add vehicle and component path specifiers used in new versions of mavlink2rest

ES-Alexander committed:
service.json: update webpage

Williangalvani committed:
Dockerfile: change base image to Python3.9-slim-bullseye

Williangalvani committed:
index.html: migrate to vue.js

Williangalvani committed:
dvl.py: change hostname to waterlinked-dvl.local by default

Williangalvani committed:
dvl.py: do not save origin in settings

Williangalvani committed:
dvl: handle the new dvl firmware changes

Williangalvani committed:
Update setting the current ROV position in the map

Williangalvani committed:
dvl: improve logging/status messages

Williangalvani committed:
dvl: automatically set rangefinder parameters

Williangalvani committed:
mavlink2resthelper: update mavlink2rest path and blueoshelper import path

Williangalvani committed:
Add dependencies offline

Williangalvani committed:
add fonts

ES-Alexander commented on dvl-a50/mavlink2resthelper.py:
```diff
@@ -166,24 +169,29 @@ def __init__(self):
 """
 
-    def get_float(self, path: str) -> float:
+    def get_float(self, path: str, vehicle: int=None, component: int=None) -> float:
```
Per @patrickelectric's comments on #6, the types here should be `Optional[int]`
```suggestion
    def get_float(self, path: str, vehicle: Optional[int]=None, component: Optional[int]=None) -> float:
```
(probably requires importing `Optional` from `typing`)

ES-Alexander commented on dvl-a50/mavlink2resthelper.py:
```diff
@@ -166,24 +169,29 @@ def __init__(self):
         return float(response)
 
-    def get(self, path: str) -> str:
+    def get(self, path: str, vehicle: int=None, component: int=None) -> str:
```
`vehicle` and `component` here are `Optional`, and we should maybe also make the return type `Optional[str]` as well, and change the failure response from `False` to `None` (or alternatively keep the return type and change the failure response to an empty string).
```suggestion
    def get(self, path: str, vehicle: Optional[int]=None, component: Optional[int]=None) -> Optional[str]:
```

ES-Alexander commented on dvl-a50/mavlink2resthelper.py:
```diff
@@ -166,24 +169,29 @@ def __init__(self):
+        vehicle_path = f"/vehicles/{vehicle}/components/{component}/messages"
+        response = request(MAVLINK2REST_URL + '/mavlink' + vehicle_path + path)
         if not response:
             return False
```
Change if changing return type to `Optional[str]`:
```suggestion
            return None
```
Otherwise need to make the return type `Union[str, bool]`, or change the failure case to return an empty string. 

Returning `False`, `None`, or `""` will all still work fine in `if not response:` checks like the one in `get_float`.

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
         self.save_settings()
 
+    def longitude_scale(self, lat):
+        scale = math.cos(lat * (math.pi / 180.0))
```
```suggestion
        scale = math.cos(math.radians(lat))
```

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+            # try old api first
+            try:
+                outputs_raw = request(
+                    "http://{0}/api/v1/outputs".format(self.hostname))
```
```suggestion
                    f"http://{self.hostname}/api/v1/outputs")
```

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                        if "port" not in output:
+                            print("no port data from API?!")
+                            self.port = 16171
+                            print("Unable to get port from API, trying to use port {self.port}")
```
```suggestion
                            print(f"Unable to get port from API, trying to use port {self.port}")
```

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                            self.port = 16171
+                            print("Unable to get port from API, trying to use port {self.port}")
+                        self.port = int(output["port"])
+                        print("Using port {0} from API".format(self.port))
```
```suggestion
                        print(f"Using port {self.port} from API")
```

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+            # then the new one
+            except:
+                port_raw = request(
+                    "http://{0}/api/v1/outputs/tcp".format(self.hostname))
```
```suggestion
                    f"http://{self.hostname}/api/v1/outputs/tcp")
```

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                        print("no port data from API?!")
+                        self.port = 16171
+                    self.port = data["port"]
+                    print("Using port {0} from API".format(self.port))
```
This is repeated from above, and could be de-dented to outside the `try`-`except` block to avoid the repetition.

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
+        lat = float(lat)
+        lon = float(lon)
+        if not self.has_origin_set():
+            self.mav.set_gps_origin(lat, lon)
+            self.origin = [lat, lon]
+            self.save_settings()
```
Should this be `self.set_gps_origin(lat, lon)` instead? Otherwise why does that function exist?

Williangalvani replied:
good catch, fixing

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
         """
         self.mav.set_gps_origin(lat, lon)
-        self.origin = [lat, lon]
+        self.origin = [float(lat), float(lon)]
```
Is this conversion necessary?
- If so, should it happen before the values get sent to over MAVLink? 
- If not, should the function signature be typed to specify floats as the expected input?

ES-Alexander reviewed (commented):
Mostly looks fine :-)
Suggestions are mostly on type hints, plus a couple of minor improvements/fixes.

Should add fixes #5 and closes #6 to the PR description.

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                    if "port" not in data:
+                        print("no port data from API?!")
+                        self.port = 16171
+                    self.port = data["port"]
```
Similarly to above, this could raise an exception if `data` has no `"port"` key
```suggestion
                    else:
                        self.port = data["port"]
```

Also, is there a reason the other one has an `int` conversion, but this one doesn't?

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                            print("no port data from API?!")
+                            self.port = 16171
+                            print("Unable to get port from API, trying to use port {self.port}")
+                        self.port = int(output["port"])
```
This could raise an exception if `output` has no `"port"` key.
```suggestion
                        else:
                            self.port = int(output["port"])
```

ES-Alexander reviewed (changes requested):
bug correction

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                        self.port = int(output["port"])
+                        print("Using port {0} from API".format(self.port))
+            # then the new one
+            except:
```
It would be nice to print what was the kind of exception.

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+            try:
+                outputs_raw = request(
+                    "http://{0}/api/v1/outputs".format(self.hostname))
+                if outputs_raw:
```
Avoid the unnecessary indentation using:
```python
if not outputs_raw:
    return
```

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                        if output["format"] != "json_v3":
+                            continue
+                        if "port" not in output:
+                            print("no port data from API?!")
```
This print is unnecessary since the following message already says the same thing.

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+                            continue
+                        if "port" not in output:
+                            print("no port data from API?!")
+                            self.port = 16171
```
This port number is used more than a single place, avoid magic numbers, create a static variable for the class as `default_port_number` to use it over the number in multiple places in the code.

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -141,15 +143,32 @@ def detect_port(self):
+            except:
+                port_raw = request(
+                    "http://{0}/api/v1/outputs/tcp".format(self.hostname))
+                if port_raw:
```
Avoid the unnecessary indentation using:
```python
if not outputs_raw:
    return
```

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
         self.should_send = should_send
         self.save_settings()
 
+    def longitude_scale(self, lat):
```
Self is unnecessary, can be a static function.

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
+        scale = math.cos(lat * (math.pi / 180.0))
+        return max(scale, 0.01)
+
+    def latLngToXY (self, lat, lon):
```
This looks unnecessary, you can use a library as `geopy` to do that for you.
```python
from geopy.distance import geodesic


origin = (30.172705, 31.526725)
dist = (30.288281, 31.732326)

print(geodesic(origin, dist).meters)  # 23576.805481751613
```
It's also really light to install under BlueOS.
![image](https://user-images.githubusercontent.com/1215497/182367003-ce1e5197-eb04-48b9-9ef5-ca69ae0f3823.png)

Also, it's not clear which unit this is returning, is it cm ? is it meters ?

Williangalvani replied:
I used exactly the same code that ardupilot uses, on purpose. I'm adding a comment of where it comes from and the units

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
+        data = json.loads(self.mav.get("/GLOBAL_POSITION_INT/message"))
+        return data['lat'] != 0 or data['lon'] != 0
+
+    def set_current_position(self, lat, lon):
```
It would be nice to have typehints or a linter in this project.

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
+            self.mav.set_gps_origin(lat, lon)
+            self.origin = [lat, lon]
+            self.save_settings()
+        else:
```
Should this else contains a print or debug message ? It does not appears to be the "default" logic while looking at the code.

Williangalvani replied:
it is the default logic, I'm adding comments to explain what is happening.

patrickelectric reviewed (changes requested):
Thanks @ES-Alexander for taking a look.

Williangalvani commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
+        y =  self.longitude_scale((lat + self.origin[0])/2) * LATLON_TO_CM * (lon-self.origin[1])
+        return [x, y]
+
+    def has_origin_set(self):
```
make sure this is working as expected.

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
+        y =  self.longitude_scale((lat + self.origin[0])/2) * LATLON_TO_CM * (lon-self.origin[1])
+        return [x, y]
+
+    def has_origin_set(self):
+        data = json.loads(self.mav.get("/GLOBAL_POSITION_INT/message"))
+        return data['lat'] != 0 or data['lon'] != 0
```
Perhaps worth a shot?
```suggestion
    def has_origin_set(self):
        # Check for record of an origin confirmation message from the autopilot
        return bool(self.mav.get("/GPS_GLOBAL_ORIGIN/message"))
```

ES-Alexander replied:
The `GPS_GLOBAL_ORIGIN` approach requires also checking that the message was received after the last autopilot restart (assuming that's even possible to check for), otherwise restarting the autopilot would require manually restarting the DVL service (or mavlink2rest).

If the `GLOBAL_POSITION_INT` approach works the way you expected then it's a more robust approach, because it would only have issues right after the autopilot restart (i.e. setting a new 'current location' wouldn't work properly because there's no GPS origin set, but the previous non-zero global position is still in mavlink2rest), but the functionality would then start working again as soon as a new `GLOBAL_POSITION_INT` message is received from the vehicle, so it's self-correcting within 1/3 of a second, which is likely fine.

ES-Alexander commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +194,44 @@ def set_should_send(self, should_send):
+            attitude = json.loads(self.mav.get("/ATTITUDE"))
+            positions = [x, y, -depth]
+            self.reset_counter += 1
+            self.mav.send_vision_position_estimate(self.timestamp, positions, reset_counter=self.reset_counter)
```
I think this should be setting `GLOBAL_VISION_POSITION_ESTIMATE`, not (local) `VISION_POSITION_ESTIMATE`.

Local should be fine for `handle_position_local` though.

ES-Alexander replied:
Noting that `handle_position_local` is incomplete at the moment, because if the user turns on `POSITION_ESTIMATE` we'll be reading the values and passing them to the autopilot, but:
- we currently aren't using the "Reset Dead Reckoning" command (it should probably happen every time we set current position, and also make use of a stored position offset that gets applied to add on the current position (as a local offset from the GPS origin) to the dead reckoning one)
- we aren't enabling DVL "Gyro Calibration" (which should presumably happen when the vehicle gets launched)
- we aren't allowing a `mounting_rotation_offset` to be specified (which should be set when mounting, if not forward-aligned with the vehicle)

~~If we're going to properly enable that, we may as well add in the additional "Configuration parameters" at the same time, which would allow the user to set speed of sound, turn off the acoustics when out of water, turn off the DVL's blinking LED, and change the range mode to a manual option.~~ **Accessible via iframe to Water Linked interface.**

If it won't be supported in this release then I'd recommend removing the `POSITION_ESTIMATE` option from the frontend, and raising an issue about it as a TODO :-)

Williangalvani replied:
yeah, it was introduced as kinda of an experimental thing that I left there, I'll play with it, but I'm leaning to hide it from end-users for now...

ES-Alexander reviewed (commented):
Some additional suggestions, given [this post](https://discuss.bluerobotics.com/t/waterlinked-dvl-a50-integration-not-working/11501/23) and [our resulting discussion](https://bluerobotics.slack.com/archives/C03SJ5EP5EG/p1659670826091929).

Williangalvani committed:
create dvlfinder

Williangalvani committed:
dvl: Find the dvl using brute-force

Williangalvani committed:
index.html: use tiles from google maps

Williangalvani committed:
dvl: use hard-coded port

Williangalvani committed:
Code style: add missing typehints

Williangalvani committed:
dvl: improve origin logic

Williangalvani committed:
static: Add static files

Williangalvani committed:
index.html: Embed waterlinked interface in an iframe

Williangalvani committed:
index.html: hide advanced options, re-organize interface

Williangalvani committed:
dvl: create debug() method

Williangalvani committed:
dvl: disable setting hostname

The discovery system is working pretty well. this only makes things
more complex

Williangalvani committed:
dvl: improve reconnection process

Williangalvani committed:
Show ip in page title

Williangalvani committed:
dvl: remove unused method

Williangalvani committed:
Emit warning in qgc when the dvl is too hot

Williangalvani committed:
Improve logging, switch to loguru

Williangalvani committed:
Use GLOBAL message for vision_position_estimate

Williangalvani committed:
deal with badly-caught exception

Williangalvani commented:
While there's more to be done/cleaned here, this has become pretty painful to work on with 31 commits. I suggest we get this in and then some more work can be done. I purposedly left some code commented as we'll not be the ones maintaining the repo long-term, and the features could be re-enabled by whoever maintains it without going trough the git history.

patrickelectric commented on Dockerfile:
```diff
@@ -1,4 +1,7 @@
+FROM python:3.9-slim-bullseye
+
+RUN apt update && apt install -y nmap
+RUN pip install python3-nmap
```
shouldn't this be in `setup.py` ?

Williangalvani replied:
true!

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +171,72 @@ def set_should_send(self, should_send):
+            if old_time == float('nan'):
+                logger.warning("Unable to read current time for GPS_GLOBAL_ORIGIN, using 0")
+        except Exception as e:
+            logger.warning(f"Unable to read current time for GPS_GLOBAL_ORIGIN, using 0: {e}")
```
as `using 0` is not supposed to set something here ?

Williangalvani replied:
damn, I just introduced this bug yesterday, nice catch

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -175,12 +171,72 @@ def set_should_send(self, should_send):
+        else:
+            logger.info("Origin has already been set, sending POSITION_ESTIMATE instead")
+            # if we already have an origin set, send a new position instead
+            x, y = self.lat_lng_to_NE_XY_cm (lat, lon)
```
there is a weird space before the parentheses :eyes: 
The project should run a linter or something to catch this kind of stuff

Williangalvani replied:
yeah, after this get in, I think I'll add black and isort at least. I'll fix this meanwhile

patrickelectric commented on dvl-a50/dvl.py:
```diff
@@ -334,26 +360,39 @@ def handle_velocity(self, data):
+            status = json.loads(request(f"http://{self.hostname}/api/v1/about/status"))
+
+            temp = float(status["temperature"])
+            if temp > 45.0:
```
It would be nice to have in the class variables for temperature_check_time and temperature_threshold

Williangalvani replied:
yep!

patrickelectric commented on dvl-a50/dvlfinder.py:
```diff
@@ -0,0 +1,50 @@
+        current_ips.append(entry["ip"])
+
+
+    scans = ['.'.join([*(ip.split('.')[0:-1]),'*']) for ip in current_ips]
```
It would be nice to have a function to encapsulate such logic, since it's not trivial to understand what is going on and the objective of such line.

patrickelectric commented on dvl-a50/dvlfinder.py:
```diff
@@ -0,0 +1,50 @@
+      for entry in network:
+        current_ips.append(entry["ip"])
+
+
```
usually double spaces between lines in python is used for function or class separation

patrickelectric commented on dvl-a50/static/js/rotatedmarker.js:
```diff
@@ -0,0 +1,57 @@
+(function() {
```
It would be nice to move to a typescript project

Williangalvani replied:
yeah, this is an external lib though.

patrickelectric reviewed (changes requested):
Just some minor questions and points

Williangalvani committed:
dvl: wait for cable-guy to come online before trying to talk to it

Williangalvani committed:
dvl: add error message in case the run loop ends

patrickelectric reviewed (approved):
Code is huge but appears to be in a good state for an initial merge

Williangalvani committed:
dvlfinder: deal with rare nmap issue

error:

192.168.2.141 - - [07/Sep/2022 17:11:59] "GET /get_status HTTP/1.1" 200 -
    output = self.run_command(scan_shlex, timeout=timeout)
  File "/usr/local/lib/python3.9/site-packages/nmap3/nmap3.py", line 264, in run_command
    raise NmapExecutionError('Error during command: "' + ' '.join(cmd) + '"\n\n' + errs.decode('utf8'))
nmap3.exceptions.NmapExecutionError: Error during command: "/usr/bin/nmap -oX - 192.168.2.* --top-ports 10 -p 80 --open"

nmap: Target.cc:503: void Target::stopTimeOutClock(const timeval*): Assertion `htn.toclock_running == true' failed.

ES-Alexander commented on dvl-a50/blueoshelper.py:
```diff
@@ -10,7 +11,7 @@ def request(url):
         return urllib.request.urlopen(url, timeout=1).read().decode()
     except Exception as error:
-        print("Error in request: {0}: {1}".format(url, error))
+        logger.warning("Error in request: {0}: {1}".format(url, error))
```
Any reason not to use f-strings for this?
```suggestion
        logger.warning(f"Error in request: {url}: {error}")
```

Williangalvani replied:
just an oversight. fixing it.

ES-Alexander commented on dvl-a50/blueoshelper.py:
```diff
@@ -29,7 +30,7 @@ def post(url, data):
-        print("Error in request: {0}: {1}".format(url, error))
-        print(error.read())
-        print(data)
+        logger.warning("Error in request: {0}: {1}".format(url, error))
+        logger.warning(error)
```
This logs the error twice, right?
```suggestion
        logger.warning(f"Error in request: {url}: {error}")
```

Williangalvani replied:
oops

ES-Alexander commented on dvl-a50/dvlfinder.py:
```diff
@@ -0,0 +1,60 @@
+    nmap = nmap3.Nmap()
+    # generate the scan mask from our current ips
+    networks = json.loads(request("http://127.0.0.1/cable-guy/v1.0/ethernet"))
+    current_networks = [network["addresses"] for network in networks]
+    # this looks like [{'ip': '192.168.2.2', 'mode': 'server'}]
+    current_ips = []
+    for network in current_networks:
+      for entry in network:
+        current_ips.append(entry["ip"])
```
Non-essential - not sure which is cleaner/clearer.
```suggestion
    current_ips = [
      entry["ip"]
      for network in networks
      for entry in network["addresses"]
    ]
```

ES-Alexander commented on dvl-a50/main.py:
```diff
@@ -92,13 +92,9 @@ def set_orientation(orientation: int):
-
-    @app.route('/origin/<lat>/<lon>')
+    @app.route('/setcurrentposition/<lat>/<lon>')
     def set_gps_origin(lat, lon):
```
```suggestion
    def set_current_position(lat, lon):
```

ES-Alexander commented on dvl-a50/mavlink2resthelper.py:
```diff
@@ -245,7 +294,20 @@ def set_param(self, param_name, param_type, param_value):
             return result.status_code == 200
         except Exception as error:
-            print("Error setting parameter: " + str(error))
+            logger.warning("Error setting parameter: " + str(error))
```
```suggestion
            logger.warning(f"Error setting parameter: {error}")
```

Williangalvani replied:
good, I also added the parameter failed to set

ES-Alexander commented on dvl-a50/mavlink2resthelper.py:
```diff
@@ -271,7 +333,7 @@ def send_vision_speed_estimate(self, speed_estimates):
 
-    def send_vision_position_estimate(self, timestamp, position_estimates, attitude_estimates=[0.0, 0.0, 0.0]):
+    def send_vision_position_estimate(self, timestamp, position_estimates, attitude_estimates=[0.0, 0.0, 0.0], reset_counter=0):
         "Sends message VISION_POSITION_ESTIMATE to flight controller"
```
It seems redundant to comment this here, and (unnecessarily) ties it to the MAVLink message in use. Probably this function should be renamed to `send_position_estimate`, and the template should be renamed to `position_estimate_template` (which are their respective purposes within the code), and then the fact that it's being sent via MAVLink's `GLOBAL_VISION_POSITION_ESTIMATE` message is isolated to where the template is defined. Same reasoning applies to the other functions (e.g. `send_vision` -> `send_position_delta`).

If we're leaving it with the current naming and comment, the comment should at least be correct :-)
```suggestion
        "Sends message GLOBAL_VISION_POSITION_ESTIMATE to flight controller"
```

Williangalvani replied:
I updated the relevant places to prepend `global_`. I understand it is a bit redundant, but I prefer to be explicit and redundant, for now...

ES-Alexander commented on dvl-a50/static/js/rotatedmarker.js:
```diff
@@ -0,0 +1,57 @@
+      this.options.rotationOrigin = this.options.rotationOrigin || iconAnchor || 'center bottom' ;
+      this.options.rotationAngle = this.options.rotationAngle || 0;
+
+      // Ensure marker keeps rotated during dragging
```
```suggestion
      // Ensure marker keeps rotating during dragging
```

Williangalvani replied:
this is external code, just included here to be available offline, I'll avoid touching it

ES-Alexander reviewed (approved):
Well done! :-)

Mostly minor suggestions that you can accept or ignore as relevant.

The `send_*` message names and docstrings are worth a think, but decision is up to you. It can also be addressed in a separate PR later if you'd prefer, especially since the current structure is already there from before this PR.

Williangalvani commented:
I'll run some tests in the pool tomorrow  and merge it then if everything looks right.
